# Lab | Data Aggregation and Filtering

In this challenge, we will continue to work with customer data from an insurance company. We will use the dataset called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by first performing data cleaning, formatting, and structuring.

1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.

2. Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign. Write your conclusions.

3. Analyze the total number of customers who have policies in each state, and then filter the results to only include states where there are more than 500 customers.

4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.

In [11]:
import pandas as pd
import numpy as np
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis.csv")

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10910 entries, 0 to 10909
Data columns (total 26 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Unnamed: 0                     10910 non-null  int64  
 1   Customer                       10910 non-null  object 
 2   State                          10279 non-null  object 
 3   Customer Lifetime Value        10910 non-null  float64
 4   Response                       10279 non-null  object 
 5   Coverage                       10910 non-null  object 
 6   Education                      10910 non-null  object 
 7   Effective To Date              10910 non-null  object 
 8   EmploymentStatus               10910 non-null  object 
 9   Gender                         10910 non-null  object 
 10  Income                         10910 non-null  int64  
 11  Location Code                  10910 non-null  object 
 12  Marital Status                 10910 non-null 

In [13]:
# Changing all colums to lowercase since is easier to work.
df.columns = df.columns.str.lower()

# Changing all colums to the same format.
df.rename(columns={'EmploymentStatus': 'Employment Status'}, inplace=True)

In [14]:
# Deleting this column because it wasn't necessary, we already have a automatic numerical column.
df = df.drop(columns='unnamed: 0')

In [15]:
df.shape

(10910, 25)

In [16]:
#Delete duplicates rows
df_drop_duplicates = df.drop_duplicates()
df = df_drop_duplicates.copy() # Save it 

In [17]:
# Check for duplicates after cleaning
duplicated_rows_after_cleaning = df.duplicated()
duplicated_rows_after_cleaning

0        False
1        False
2        False
3        False
4        False
         ...  
10905    False
10906    False
10907    False
10908    False
10909    False
Length: 10873, dtype: bool

In [18]:
df.shape # To see more cleary how many we have left 

(10873, 25)

In [19]:
num_nans = df.isna().sum() # Check how many nulls values do we have.
print(num_nans)

customer                            0
state                             614
customer lifetime value             0
response                          614
coverage                            0
education                           0
effective to date                   0
employmentstatus                    0
gender                              0
income                              0
location code                       0
marital status                      0
monthly premium auto                0
months since last claim           623
months since policy inception       0
number of open complaints         623
number of policies                  0
policy type                         0
policy                              0
renew offer type                    0
sales channel                       0
total claim amount                  0
vehicle class                     608
vehicle size                      608
vehicle type                     5465
dtype: int64


`1. Create a new DataFrame that only includes customers who have a total_claim_amount greater than $1,000 and have a response of "Yes" to the last marketing campaign.`

In [22]:
filtered_df = df[(df['total claim amount'] > 1000) & (df['response'] == 'Yes')]

In [23]:
filtered_df

customer       state  customer lifetime value response  coverage  \
189    OK31456  California             11009.130490      Yes   Premium   
236    YJ16163      Oregon             11009.130490      Yes   Premium   
419    GW43195      Oregon             25807.063000      Yes  Extended   
442    IP94270     Arizona             13736.132500      Yes   Premium   
587    FJ28407  California              5619.689084      Yes   Premium   
...        ...         ...                      ...      ...       ...   
10351  FN44127      Oregon              3508.569533      Yes  Extended   
10373  XZ64172      Oregon             10963.957230      Yes   Premium   
10487  IX60941      Oregon              3508.569533      Yes  Extended   
10565  QO62792      Oregon              7840.165778      Yes  Extended   
10708  CK39096      Oregon              5619.689084      Yes   Premium   

                  education effective to date employmentstatus gender  income  \
189                Bachelor           1/24/11         Employed      F   51643   
236                Bachelor           1/24/11         Employed      F   51643   
419                 College           2/13/11         Employed      F   71210   
442                  Master           2/13/11         Disabled      F   16181   
587    High School or Below           1/26/11       Unemployed      M       0   
...                     ...               ...              ...    ...     ...   
10351               College            1/5/11    Medical Leave      M   20978   
10373  High School or Below            2/8/11         Employed      M   55687   
10487               College            1/5/11    Medical Leave      M   20978   
10565               College           1/14/11         Employed      M   58414   
10708  High School or Below           1/26/11       Unemployed      M       0   

       ... number of open complaints number of policies     policy type  \
189    ...                       0.0                  1  Corporate Auto   
236    ...                       0.0                  1    Special Auto   
419    ...                       1.0                  2   Personal Auto   
442    ...                       0.0                  8   Personal Auto   
587    ...                       0.0                  1   Personal Auto   
...    ...                       ...                ...             ...   
10351  ...                       1.0                  1   Personal Auto   
10373  ...                       0.0                  1  Corporate Auto   
10487  ...                       1.0                  1   Personal Auto   
10565  ...                       2.0                  1   Personal Auto   
10708  ...                       0.0                  1   Personal Auto   

             policy  renew offer type  sales channel  total claim amount  \
189    Corporate L3            Offer2          Agent         1358.400000   
236      Special L3            Offer2          Agent         1358.400000   
419     Personal L2            Offer1         Branch         1027.200000   
442     Personal L2            Offer1            Web         1261.319869   
587     Personal L1            Offer2            Web         1027.000029   
...             ...               ...            ...                 ...   
10351   Personal L2            Offer2         Branch         1176.278800   
10373  Corporate L2            Offer1          Agent         1324.800000   
10487   Personal L3            Offer2         Branch         1176.278800   
10565   Personal L3            Offer2          Agent         1008.000000   
10708   Personal L3            Offer2            Web         1027.000029   

       vehicle class vehicle size vehicle type  
189       Luxury Car      Medsize          NaN  
236       Luxury Car      Medsize            A  
419       Luxury Car        Small            A  
442              SUV      Medsize            A  
587              SUV      Medsize            A  
...              ...          ...   

In [21]:
filtered_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 67 entries, 189 to 10708
Data columns (total 25 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   customer                       67 non-null     object 
 1   state                          67 non-null     object 
 2   customer lifetime value        67 non-null     float64
 3   response                       67 non-null     object 
 4   coverage                       67 non-null     object 
 5   education                      67 non-null     object 
 6   effective to date              67 non-null     object 
 7   employmentstatus               67 non-null     object 
 8   gender                         67 non-null     object 
 9   income                         67 non-null     int64  
 10  location code                  67 non-null     object 
 11  marital status                 67 non-null     object 
 12  monthly premium auto           67 non-null     int64

`2.Using the original Dataframe, analyze the average total_claim_amount by each policy type and gender for customers who have responded "Yes" to the last marketing campaign.`

In [24]:
# Group by policy type and gender, then calculate the average total_claim_amount
avg_claim_amount = df[df['response'] == 'Yes'].groupby(['policy type', 'gender'])[['total claim amount']].mean().round(2)
avg_claim_amount


total claim amount
policy type    gender                    
Corporate Auto F                   433.74
               M                   407.83
Personal Auto  F                   452.76
               M                   457.01
Special Auto   F                   453.28
               M                   429.53

**Write your conclusions.**

After analyzing the data grouped by policy type and gender, it's evident that there are variations in the average total claim amount across different policy types and genders.

For Corporate Auto policies, females have an average total claim amount of 433.74, while males have a slightly lower average of 407.83. This indicates that, on average, female policyholders in the Corporate Auto category tend to file claims for higher amounts compared to male policyholders.

In the Personal Auto category, the situation is reversed, with females having a slightly higher average total claim amount of 452.76 compared to males with an average of 457.01. However, the difference in averages is minimal in this case.

Lastly, in the Special Auto category, females have an average total claim amount of 453.28, which is slightly higher than the average for males at 429.53.

In conclusion, while there are some variations in the average total claim amounts across different policy types and genders, the differences are not substantial. However, these insights can still be valuable for insurance companies to understand their claims patterns and potentially tailor their services or pricing strategies accordingly.


`3.Analyze the total number of customers who have policies in each state, and then filter the results to only 
include states where there are more than 500 customers.`


In [25]:
policy_state = df.groupby('state')[['policy']].count()
policy_state

policy
state             
Arizona       1934
California    3548
Nevada         992
Oregon        2897
Washington     888

In [26]:
# They all have more than 500 customers. 

#states_over_500 = policy_state[policy_state > 500]

`4. Find the maximum, minimum, and median customer lifetime value by education level and gender. Write your conclusions.`

In [27]:
stats_by_education_gender = df.groupby(['education', 'gender'])['customer lifetime value'].agg(['max', 'min', 'median']).round(2)
stats_by_education_gender

max      min   median
education            gender                            
Bachelor             F       73225.96  1904.00  5632.61
                     M       67907.27  1898.01  5548.03
College              F       61850.19  1898.68  5623.61
                     M       61134.68  1918.12  6005.85
Doctor               F       44856.11  2395.57  5332.46
                     M       32677.34  2267.60  5577.67
High School or Below F       55277.45  2144.92  6035.09
                     M       83325.38  1940.98  6286.73
Master               F       51016.07  2417.78  5729.86
                     M       50568.26  2272.31  5579.10

**Conclusions**

Upon analyzing the data on income grouped by education level and gender, several observations can be made regarding the maximum, minimum, and median income values.

For individuals with a Bachelor's degree, the maximum income is highest for males at 67,907.27, whereas females have a maximum income of 73,225.96. However, both genders have similar minimum and median incomes, suggesting a relatively balanced distribution of income within this education level.

Similarly, individuals with a College degree also exhibit a balanced distribution of income between genders, with slightly higher maximum income for females compared to males. However, the differences in minimum and median income between genders are minimal.

For those with a Doctorate degree, females have a higher maximum income compared to males, but the median income for both genders is relatively close. The minimum income for males is lower compared to females, indicating potential disparities in income distribution within this education level.

Individuals with a High School diploma or below demonstrate significant differences in income distribution between genders. Males have notably higher maximum income compared to females, indicating potential gender disparities in earnings within this education category. Additionally, the median income for males is higher than for females, further highlighting income inequalities.

Lastly, individuals with a Master's degree exhibit relatively similar income distributions between genders, with minor differences in maximum and median income values. However, males have a slightly lower minimum income compared to females within this education level.

In conclusion, while there are variations in income distribution across different education levels and genders, disparities are more pronounced among individuals with lower levels of education, particularly in the High School or Below category. These findings underscore the importance of addressing gender disparities in income, particularly among individuals with lower levels of education.


## Bonus

5. The marketing team wants to analyze the number of policies sold by state and month. Present the data in a table where the months are arranged as columns and the states are arranged as rows.

6.  Display a new DataFrame that contains the number of policies sold by month, by state, for the top 3 states with the highest number of policies sold.

*Hint:*
- *To accomplish this, you will first need to group the data by state and month, then count the number of policies sold for each group. Afterwards, you will need to sort the data by the count of policies sold in descending order.*
- *Next, you will select the top 3 states with the highest number of policies sold.*
- *Finally, you will create a new DataFrame that contains the number of policies sold by month for each of the top 3 states.*

7. The marketing team wants to analyze the effect of different marketing channels on the customer response rate.

Hint: You can use melt to unpivot the data and create a table that shows the customer response rate (those who responded "Yes") by marketing channel.

External Resources for Data Filtering: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9